app/uk/gov/hmrc/vulnerabilities/persistence/VulnerabilitySummariesRepository.scala

```diff
@@ -60,15 +61,41 @@ class VulnerabilitySummariesRepository @Inject()(
     collection.find().toFuture()
 
   private implicit val tc = TransactionConfiguration.strict
-
-  def deleteOldAndInsertNewSummaries(summaries: Seq[VulnerabilitySummary]): Future[Int] =
+  def deleteOldAndInsertNewSummaries(newSummaries: Seq[VulnerabilitySummary], allSvDeps: Seq[ServiceVersionDeployments]): Future[Int] =
     withSessionAndTransaction(session =>
       for {
-        _        <- collection.deleteMany(session, Filters.empty()).toFuture()
-        inserted <- collection.insertMany(session, summaries).toFuture().map(_.getInsertedIds.size())
+        currentSummaries         <- getVulnerabilitySummaries()
+        filteredCurrentSummaries =  filterOutEnvironmentsNoLongerDeployedTo(currentSummaries, allSvDeps)
+        mergedSummaries          =  ensureOldVulnerabilitiesStillIncluded(filteredCurrentSummaries, newSummaries)
+        _                        <- collection.deleteMany(session, Filters.empty()).toFuture()
+        inserted                 <- collection.insertMany(session, mergedSummaries).toFuture().map(_.getInsertedIds.size())
       } yield inserted
     )
 
+  // Removes occurrences where they are no longer in ServiceVersionDeployments
+  def filterOutEnvironmentsNoLongerDeployedTo(currentSummaries: Seq[VulnerabilitySummary], allSvDeps: Seq[ServiceVersionDeployments]): Seq[VulnerabilitySummary] = {
+    currentSummaries
+      .map(vs => vs.copy(occurrences = vs.occurrences
+        .map(occ => occ.copy(envs = occ.envs.filter(env => allSvDeps.exists(svd => svd.serviceName === occ.service && svd.version === occ.serviceVersion && svd.environments.contains(env)))))
+        .filter(occ => occ.envs.nonEmpty)))
+      .filter(vs => vs.occurrences.nonEmpty)
+  }
+
+  // If a vulnerability only exists in filteredCurrentSummaries add it back to newSummaries. ditto for environments
+  def ensureOldVulnerabilitiesStillIncluded(filteredCurrentSummaries: Seq[VulnerabilitySummary], newSummaries: Seq[VulnerabilitySummary]): Seq[VulnerabilitySummary] = {
+    val summariesNotInNew = filteredCurrentSummaries.filterNot(cur => newSummaries.exists(newS => newS.distinctVulnerability.id === cur.distinctVulnerability.id))
+    val newSummariesWithEnvironmentsNotInNew = newSummaries.map(ns => {
+      filteredCurrentSummaries.find(summary => summary.distinctVulnerability.id === ns.distinctVulnerability.id) match {
+        case Some(cs) => ns.copy(occurrences = ns.occurrences.map(occ => cs.occurrences.find(x => x.service == occ.service && x.serviceVersion === occ.serviceVersion && x.envs =!= occ.envs) match {
+          case Some(value) => value.copy(envs = occ.envs.appendedAll(value.envs).distinct)
+          case None => occ
+        }))
+        case None => ns
+      }
+    })
```

Comment:
Could we use `fold` here (or map with getOrElse)
```suggestion
    val newSummariesWithEnvironmentsNotInNew = newSummaries.map(ns => {
      filteredCurrentSummaries.find(summary => summary.distinctVulnerability.id === ns.distinctVulnerability.id)fold(ns)(cd  => ns.copy(occurrences = ns.occurrences.map(occ => cs.occurrences.find(x => x.service == occ.service && x.serviceVersion === occ.serviceVersion && x.envs =!= occ.envs).fold(occ)(_.copy(envs = occ.envs.appendedAll(value.envs).distinct))
        }))
      }
    })
```